### PR TITLE
Anvil Documentation

### LINKED ISSUE #190: Document `forge node`
When merged, we should document `forge node`

See https://github.com/gakonst/foundry/pull/1037

### PR DESCRIPTION
**Pull request by Leutwiler**
I wrote four topics for the Anvil documentation:
1) Anvil Flags: covers the possible Anvil flags
2) Mining Modes: talks about the mining modes
3) Supported Transport Layers: defines the supported transport layers (as far as I know and based on a conversation with Matthias, WS is still not supported and requires some code rewriting, so I documented that it will be supported soon)
4) Supported RPC Methods: covers every Anvil's RPC method

This documentation was an incredible experience for me! I could learn many new things about Foundry and other general subjects.

Huge thanks to @mattsse ! He was crucial for this documentation because he helped me a lot with some questions I had! :)

Fix #190

### CODE REVIEW

File: src/anvil/rpcMethods.md [hunk 1]
```diff
@@ -0,0 +1,170 @@
+## Supported RPC Methods
+### Standard Methods
```
Comment (mattsse) [inline src/anvil/rpcMethods.md:2]: can we link to the eth wiki here https://eth.wiki/json-rpc/API

File: src/anvil/rpcMethods.md [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+Sets the base fee of the next block
+
+### Special Methods
+`evm_setAutomine`  
+Enables or disables, based on the single boolean argument, the automatic mining of new blocks with each new transaction submitted to the network.
+  
+`evm_setIntervalMining`  
+Sets the mining behavior to interval with the given interval (seconds)
+
+`evm_snapshot`
```
Comment (mattsse) [inline src/anvil/rpcMethods.md:142]: these are all coming from ganache we should link here to https://github.com/trufflesuite/ganache-cli-archive/blob/master/README.md

File: src/anvil/rpcMethods.md [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+`eth_sendUnsignedTransaction`  
+Execute a transaction regardless of signature status
+
+`txpool_status`  
+Returns the number of transactions currently pending for inclusion in the next block(s), as well as the ones that are being scheduled for future execution only  
+
+`txpool_inspect`  
+Returns a summary of all the transactions currently pending for inclusion in the next block(s), as well as the ones that are being scheduled for future execution only
+
+`txpool_content`
```
Comment (mattsse) [inline src/anvil/rpcMethods.md:169]: these are geth endpoints, we should link here to https://geth.ethereum.org/docs/rpc/ns-txpool

File: src/anvil/transportLayers.md [hunk 4]
```diff
@@ -0,0 +1,6 @@
+## Supported Transport Layers
+HTTP is already supported and WebSocket (WS) will be supported soon. The server listens on port 8545 by default, but it can be changed by running the following command:
+
+```
+anvil --port <PORT>
+```
```
Comment (mattsse) [inline src/anvil/transportLayers.md:6]: I don't think this needs to be a separate file

File: src/anvil/rpcMethods.md [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+  
+`eth_feeHistory`  
+
+`debug_traceTransaction`  
+  
+`trace_transaction`
+  
+`trace_block`
+
+### Custom Methods
```
Comment (mattsse) [inline src/anvil/rpcMethods.md:89]: need to add info that the `anvil` namespace is an alias for `hardhat`

**Review by mattsse — CHANGES_REQUESTED** [review]
thanks so much for this, this is great!

couple nits otherwise this is a great start

File: src/SUMMARY.md [hunk 6]
```diff
@@ -46,6 +46,10 @@
 # Anvil Overview 
 
 - [Overview of Anvil](./anvil/README.md)
+- [Anvil Flags](./anvil/AnvilFlags.md)
+- [Mining Modes](./anvil/MiningModes.md)
+- [Supported Transport Layers](./anvil/transportLayers.md)
+- [Supported RPC Methods](./anvil/rpcMethods.md)
```
Comment (onbjerg) [inline src/SUMMARY.md:52]: All of these files should be named in line with the other files using lowercase and with `-` as a separator

File: src/anvil/AnvilFlags.md [hunk 7]
```diff
@@ -0,0 +1,129 @@
+## Anvil Flags
```
Comment (onbjerg) [inline src/anvil/AnvilFlags.md:1]: This is better suited for a command reference like Forge and Cast, and should live in `src/reference` as well as follow the general structure that Cast and Forge does

File: src/anvil/AnvilFlags.md [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+`--sender <ADDRESS> `  
+&nbsp;&nbsp;&nbsp;&nbsp; The address which will be executing tests
+
+`-v, --verbosity`  
+&nbsp;&nbsp;&nbsp;&nbsp; Verbosity of the EVM. Pass multiple times to increase the verbosity (e.g. -v, -vv, -vvv)  
+ &nbsp;&nbsp;&nbsp;&nbsp; Verbosity levels:  
+            &nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - 2: Print logs for all tests  
+            &nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - 3: Print execution traces for failing tests  
+            &nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - 4: Print execution traces for all tests, and setup traces for failing tests  
+            &nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - 5: Print execution and setup traces for all tests
```
Comment (onbjerg) [inline src/anvil/AnvilFlags.md:71]:
```suggestion
 &nbsp;&nbsp;&nbsp;&nbsp; Verbosity levels:  
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - 2: Print logs for all tests  
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - 3: Print execution traces for failing tests  
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - 4: Print execution traces for all tests, and setup traces for failing tests  
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; - 5: Print execution and setup traces for all tests
```

File: src/anvil/AnvilFlags.md [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+`--no-cors`  
+&nbsp;&nbsp;&nbsp;&nbsp; Disable CORS
+
+
+### Examples
+
+1. Set the number of accounts to 15 and their balance to 300 ETH
+```
+anvil --accounts 15 --balance 300
+```
```
Comment (onbjerg) [inline src/anvil/AnvilFlags.md:119]:
````suggestion
  ```sh
  anvil --accounts 15 --balance 300
  ```
````

File: src/anvil/AnvilFlags.md [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+
+2. Choose the address which will execute the tests
+```
+anvil --sender 0xC8479C45EE87E0B437c09d3b8FE8ED14ccDa825E
+```
```
Comment (onbjerg) [inline src/anvil/AnvilFlags.md:124]:
````suggestion
  ```sh
  anvil --sender 0xC8479C45EE87E0B437c09d3b8FE8ED14ccDa825E
  ```
````

File: src/anvil/AnvilFlags.md [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+
+3. Change how transactions are sorted in the mempool to FIFO
+```
+anvil --order fifo
+```
```
Comment (onbjerg) [inline src/anvil/AnvilFlags.md:129]:
````suggestion
  ```sh
  anvil --order fifo
  ```
````

File: src/anvil/MiningModes.md [hunk 12]
```diff
@@ -0,0 +1,14 @@
+## Mining Modes
+Mining modes refer to how frequent blocks are mined using Anvil. By default, it automatically generates a new block as soon as a transaction is submitted.
+
+You can change this setting to interval mining if you will, which means that a new block will be generated in a given period of time selected by the user. If you want to go for this type of mining, you can do it by adding the `--block-time <block-time-in-seconds>` flag, like in the following example.
+```
+// Produces a new block every 10 seconds
+anvil --block-time 10
+```
```
Comment (onbjerg) [inline src/anvil/MiningModes.md:8]:
````suggestion
```sh
# Produces a new block every 10 seconds
anvil --block-time 10
```
````

File: src/anvil/MiningModes.md [hunk 13]
```diff
@@ -0,0 +1,14 @@
...
+
+There's also a third mining mode called never. In this case, it disables auto and interval mining, and mine on demand instead. You can do this by typing:
+```
+// Enables never mining mode
+anvil --no-mining
+```
```
Comment (onbjerg) [inline src/anvil/MiningModes.md:14]:
````suggestion
```sh
# Enables never mining mode
anvil --no-mining
```
````

File: src/anvil/rpcMethods.md [hunk 14]
```diff
@@ -0,0 +1,178 @@
+## Supported RPC Methods
+### Standard Methods
+The standard methods are based on [this](https://eth.wiki/json-rpc/API) reference.
+
+`web3_clientVersion`
```
Comment (onbjerg) [inline src/anvil/rpcMethods.md:5]: Please put this in a list

File: src/anvil/rpcMethods.md [hunk 15]
```diff
@@ -0,0 +1,178 @@
...
+`eth_feeHistory`  
+
+`debug_traceTransaction`  
+  
+`trace_transaction`
+  
+`trace_block`
+
+### Custom Methods
+`Anvil` namespace is an alias for `hardhat`, just like in [this](https://hardhat.org/hardhat-network/reference#hardhat-network-methods) Hardhat documentation.
```
Comment (onbjerg) [inline src/anvil/rpcMethods.md:92]:
```suggestion
The `anvil_*` namespace is an alias for `hardhat`. For more info, refer to the [Hardhat documentation](https://hardhat.org/hardhat-network/reference#hardhat-network-methods).
```

**Review by onbjerg — CHANGES_REQUESTED** [review]
Some structural comments:

- All of this should probably live in a command reference like Forge and Cast have, since this is reference documentation and not really guides. The place where this content was put is more geared towards the latter.
- Almost all of these can be merged into one overall Forge reference document. I also don't think we need to list out every standard JSONRPC method we support, only the differences/Anvil specific ones

Thanks for taking a stab at this :smile:

File: src/anvil/rpcMethods.md (hunk 14 above)
Comment (Leutwiler) [reply src/anvil/rpcMethods.md]: The new version has the standard methods on a listed format. I don't think it looks good for the custom and special methods, as they have an explanation underneath them, but if you think it's necessary to also list them, I can do it.

What's your opinion about it?

**Comment by Leutwiler** [comment]
> * All of this should probably live in a command reference like Forge and Cast have, since this is reference documentation and not really guides. The place where this content was put is more geared towards the latter.
> * Almost all of these can be merged into one overall Forge reference document. I also don't think we need to list out every standard JSONRPC method we support, only the differences/Anvil specific ones

I got a little bit lost with the "all of this" and "almost all of these". As far as I could understand, I changed the Anvil Flags part to References and merged the other files in one, which I'm gonna keep in the same place.

I really think we should keep with the standard RPC methods. From a beginner's perspective, I didn't know what were RPC methods and which were the standard methods before I started to write this documentation, so I think this part may help new people who are not familiar with this topic. Hardhat's documentation also covers their available standard RPC methods in the same format, that's also why I decided to keep them. What do you think about that?

> * The file names should follow the structure of pre-existing files (lowercase and `-` as a seperator)
This and all the other comments you wrote are already done! :)

File: src/SUMMARY.md [hunk 16]
```diff
@@ -178,6 +179,11 @@
       - [cast wallet sign](./reference/cast/cast-wallet-sign.md)
       - [cast wallet vanity](./reference/cast/cast-wallet-vanity.md)
       - [cast wallet verify](./reference/cast/cast-wallet-verify.md)
+  - [`anvil` Flags](./reference/forge/README.md)
+    - [Option Flags](./reference/anvil/anvil-options.md)
+    - [EVM Options Flags](./reference/anvil/anvil-evm-options.md)
+    - [Executor Environment Config Flags](./reference/anvil/anvil-environment.md)
+    - [Server Options Flags](./reference/anvil/anvil-server.md)
```
Comment (onbjerg) [inline src/SUMMARY.md:186]: This should just be 

```suggestion
  - [`anvil` Reference](./reference/anvil/README.md)
```

And all of the flags and an overview should be in the README, just like we have for Cast and Forge

File: src/reference/README.md [hunk 17]
```diff
@@ -2,6 +2,7 @@
 
 - [forge Commands](./forge/)
 - [cast Commands](./cast/)
+- [anvil Flags](./anvil/)
```
Comment (onbjerg) [inline src/reference/README.md:5]:
```suggestion
- [anvil Reference](./anvil/)
```

File: src/reference/anvil/README.md [hunk 18]
```diff
@@ -0,0 +1,6 @@
+## anvil Flags
```
Comment (onbjerg) [inline src/reference/anvil/README.md:1]:
```suggestion
## anvil Reference
```

**Review by onbjerg — CHANGES_REQUESTED** [review]
Still a bit to go - the structure should follow that of Forge and Cast in the reference :smile: 

For an idea of what I mean, see https://github.com/foundry-rs/book/pull/359

Additionally, mining modes and RPC methods could also be in the description of the commnad reference - see for example the longer-form description of forge build

**Comment by Leutwiler** [comment]
I think everything is fixed now! If there are any other issues, just let me know! 😄

File: src/reference/anvil/README.md [hunk 19]
```diff
@@ -0,0 +1,343 @@
...
+There's also a third mining mode called never. In this case, it disables auto and interval mining, and mine on demand instead. You can do this by typing:
+```sh
+# Enables never mining mode
+anvil --no-mining
+```
+
+#### Supported Transport Layers
+HTTP is already supported and WebSocket (WS) will be supported soon. The server listens on port 8545 by default, but it can be changed by running the following command:
+
+```
```
Comment (onbjerg) [inline src/reference/anvil/README.md:35]:
````suggestion
```sh
````

**Review by onbjerg — APPROVED** [review]
Thanks for this and sorry about all the nits! :smile:

**Comment by mattsse** [comment]
thanks so much @Leutwiler

**Pull request merged**